File: easynav/src/EasyNavigation/easynav_support_py/easynav_goalmanager_py/goal_manager_client.py
# Copyright 2025 Intelligent Robotics Lab
#
# This file is part of the project Easy Navigation (EasyNav in short)
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

from __future__ import annotations

from enum import Enum

# Interfaces
from easynav_interfaces.msg import NavigationControl
from geometry_msgs.msg import PoseStamped
from nav_msgs.msg import Goals

from rclpy.node import Node
from rclpy.qos import QoSProfile


class ClientState(Enum):
    IDLE = 0
    SENT_GOAL = 1
    SENT_PREEMPT = 2
    ACCEPTED_AND_NAVIGATING = 3
    NAVIGATION_FINISHED = 4
    NAVIGATION_REJECTED = 5
    NAVIGATION_FAILED = 6
    NAVIGATION_CANCELLED = 7
    ERROR = 8


class GoalManagerClient:
    """Python client compatible with C++ GoalManagerClient."""

    def __init__(
        self,
        node: Node
    ) -> None:
        self.node = node

        self.id = self.node.get_name() + '_goal_manager_client'
        self.id = self.node.get_name() + '_goal_manager_client'

        self.control_topic = 'easynav_control'
        self.goal_topic = 'goal_pose'

        self.state = ClientState.IDLE

        self.last_control = NavigationControl()
        self.last_feedback = NavigationControl()
        self.last_result = NavigationControl()

        qos = QoSProfile(depth=100)
        self._control_pub = self.node.create_publisher(NavigationControl, self.control_topic, qos)
        self._goal_pub = self.node.create_publisher(PoseStamped, self.goal_topic, qos)
        self._control_sub = self.node.create_subscription(
            NavigationControl, self.control_topic, self._on_control, qos
        )

    def send_goal(self, goal: PoseStamped) -> None:
        """Send a single goal to the GoalManager."""
        self.node.get_logger().info('Sending navigation goal')

        goals = Goals()
        goals.header = goal.header
        goals.goals.append(goal)

        self.send_goals(goals)

    def send_goals(self, goals: Goals) -> None:
        """Send a list of goals to the GoalManager."""
        if len(goals.goals) == 0:
            self.node.get_logger().error('Trying to command empty goals')
            return

        msg = NavigationControl()

        match self.state:
            case ClientState.IDLE:
                self.state = ClientState.SENT_GOAL
                msg.type = NavigationControl.REQUEST
            case ClientState.ACCEPTED_AND_NAVIGATING:
                self.state = ClientState.SENT_PREEMPT
                msg.type = NavigationControl.REQUEST
            case _:
                self.node.get_logger().error(
                    f'Trying to send new goals in state {self.state.name}. Ignoring')
                return

        msg.header = goals.header
        msg.user_id = self.id
        msg.seq = self.last_control.seq + 1
        msg.goals = goals

        self._control_pub.publish(msg)

    def cancel(self) -> None:
        """Cancel the current goal."""
        self.node.get_logger().info('Sending nevigation cancelation')

        if self.state != ClientState.ACCEPTED_AND_NAVIGATING:
            self.node.get_logger().error(
                f'Triying to cancel a non-active navigation (state {self.state.name})')
            return

        msg = NavigationControl()
        msg.type = NavigationControl.CANCEL
        msg.header = self.last_control.header
        msg.user_id = self.id
        msg.seq = self.last_control.seq + 1

        self.node.get_logger().debug('Navigation cancelation sent')

        self._control_pub.publish(msg)

    def reset(self) -> None:
        """Reset internal client state."""
        if (
            self.state == ClientState.NAVIGATION_FINISHED or
            self.state == ClientState.NAVIGATION_REJECTED or
            self.state == ClientState.NAVIGATION_FAILED or
            self.state == ClientState.NAVIGATION_CANCELLED or
            self.state == ClientState.ERROR
        ):
            self.state = ClientState.IDLE
        else:
            self.node.get_logger().error(
                f'Triying to reset navigation in a a non-finished nav state {self.state.name}')

    def get_state(self) -> ClientState:
        """Get the current internal state."""
        return self.state

    def get_last_control(self) -> NavigationControl:
        """Get the last control message sent or received."""
        return self.last_control

    def get_feedback(self) -> NavigationControl:
        """Get the most recent feedback received."""
        return self.last_feedback

    def get_result(self) -> NavigationControl:
        """Get the last result message received."""
        return self.last_result

    def _on_control(self, msg: NavigationControl) -> None:
        if msg.user_id == self.id:  # Avoid self messages
            return
        if msg.nav_current_user_id != self.id:  # Avoid messages to others
            return

        self.node.get_logger().debug(
            f'Received a navigation {msg.type} msg with user_id {msg.user_id}')

        if (
            self.state != ClientState.IDLE and
            self.state != ClientState.NAVIGATION_FINISHED and
            self.state != ClientState.NAVIGATION_FAILED and
            self.state != ClientState.NAVIGATION_CANCELLED and
            self.state != ClientState.ERROR
        ):
            match self.state:
                case ClientState.SENT_GOAL:
                    match msg.type:
                        case NavigationControl.FEEDBACK:
                            self.node.get_logger().debug(
                                'Getting navigation feedback while waiting for acceptance')
                            self.last_feedback = msg
                        case NavigationControl.ACCEPT:
                            self.node.get_logger().debug('Goal accepted. Navigating')
                            self.state = ClientState.ACCEPTED_AND_NAVIGATING
                        case NavigationControl.REJECT:
                            self.node.get_logger().error('Rejected navigation goal')
                            self.state = ClientState.NAVIGATION_FAILED
                        case NavigationControl.ERROR:
                            self.node.get_logger().error('Error in navigation')
                            self.state = ClientState.ERROR
                        case _:
                            self.node.get_logger().error(
                                'State SENT_GOAL; Unexpected message: "%d": "%s"' %
                                (msg.type, msg.status_message))
                            self.state = ClientState.ERROR
                case ClientState.SENT_PREEMPT:
                    match msg.type:
                        case NavigationControl.FEEDBACK:
                            self.node.get_logger().debug('Getting navigation feedback')
                            self.last_feedback = msg
                        case NavigationControl.ACCEPT:
                            self.node.get_logger().debug('Goal preemption accepted. Navigating')
                            self.state = ClientState.ACCEPTED_AND_NAVIGATING
                        case NavigationControl.REJECT:
                            self.node.get_logger().error('Rejected navigation preempt goal')
                            self.state = ClientState.ACCEPTED_AND_NAVIGATING
                        case NavigationControl.ERROR:
                            self.node.get_logger().error('Error in preempting navigation')
                            self.state = ClientState.ERROR
                        case _:
                            self.node.get_logger().error(
                                'State SENT_PREEMPT; Unexpected message: "%d": "%s"' %
                                (msg.type, msg.status_message))
                            self.state = ClientState.ERROR
                case ClientState.ACCEPTED_AND_NAVIGATING:
                    match msg.type:
                        case NavigationControl.FEEDBACK:
                            self.node.get_logger().debug('Getting navigation feedback')
                            self.last_feedback = msg
                        case NavigationControl.FINISHED:
                            self.node.get_logger().info('Navigation succesfully finished')
                            self.last_result = msg
                            self.state = ClientState.NAVIGATION_FINISHED
                        case NavigationControl.FAILED:
                            self.node.get_logger().error(
                                'Navigation with error finished: %s"' % msg.status_message)
                            self.last_result = msg
                            self.state = ClientState.NAVIGATION_FAILED
                        case NavigationControl.CANCELLED:
                            self.node.get_logger().error('Navigation cancelled')
                            self.last_result = msg
                            self.state = ClientState.NAVIGATION_CANCELLED
                        case _:
                            self.node.get_logger().error(
                                'State ACCEPTED_AND_NAVIGATING; Unexpected message: "%d": "%s"' %
                                msg.type, msg.status_message)
                            self.last_result = msg
                            self.state = ClientState.ERROR
                case _:
                    self.node.get_logger().error(f'State not managed: {self.state.name}')
                    self.state = ClientState.ERROR

        self.last_control = msg

File: easynav/src/EasyNavigation/easynav_system/src/easynav_system/GoalManagerClient.cpp
// Copyright 2025 Intelligent Robotics Lab
//
// This file is part of the project Easy Navigation (EasyNav in short)
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

/// \file
/// \brief Implementation of the GoalManager class.


#include "easynav_interfaces/msg/navigation_control.hpp"
#include "geometry_msgs/msg/pose_stamped.hpp"
#include "nav_msgs/msg/goals.hpp"

#include "easynav_system/GoalManagerClient.hpp"

namespace easynav
{

using namespace  std::placeholders;


GoalManagerClient::GoalManagerClient(rclcpp::Node::SharedPtr node)
: node_(node)
{
  control_sub_ = node_->create_subscription<easynav_interfaces::msg::NavigationControl>(
    "easynav_control", 100,
    std::bind(&GoalManagerClient::control_callback, this, _1)
  );

  control_pub_ = node_->create_publisher<easynav_interfaces::msg::NavigationControl>(
    "easynav_control", 100);

  id_ = std::string(node_->get_name()) + "_goal_manager_client";
  state_ = State::IDLE;
}

void
GoalManagerClient::control_callback(easynav_interfaces::msg::NavigationControl::UniquePtr msg)
{
  if (msg->user_id == id_) {return;}  // Avoid self messages
  if (msg->nav_current_user_id != id_) {return;}  // Avoid messages to others

  RCLCPP_DEBUG(node_->get_logger(), "Received a navigation %d msg with user_id %s",
    msg->type, msg->user_id.c_str());

  switch (state_) {
    case State::IDLE:
    case State::NAVIGATION_FINISHED:
    case State::NAVIGATION_FAILED:
    case State::NAVIGATION_CANCELLED:
    case State::ERROR:
      break;
    case State::SENT_GOAL:
      switch (msg->type) {
        case easynav_interfaces::msg::NavigationControl::FEEDBACK:
          RCLCPP_DEBUG(node_->get_logger(),
          "Getting navigation feedback while waiting for acceptance");
          last_feedback_ = *msg;
          break;
        case easynav_interfaces::msg::NavigationControl::ACCEPT:
          RCLCPP_DEBUG(node_->get_logger(), "Goal accepted. Navigating");
          state_ = State::ACCEPTED_AND_NAVIGATING;
          break;
        case easynav_interfaces::msg::NavigationControl::REJECT:
          RCLCPP_ERROR(node_->get_logger(), "Rejected navigation goal");
          state_ = State::NAVIGATION_FAILED;
          break;
        case easynav_interfaces::msg::NavigationControl::ERROR:
          RCLCPP_ERROR(node_->get_logger(), "Error in navigation");
          state_ = State::ERROR;
          break;
        default:
          RCLCPP_ERROR(
            node_->get_logger(), "State SENT_GOAL; Unexpected message : %d: %s",
            msg->type, msg->status_message.c_str());
          state_ = State::ERROR;
      }
      break;
    case State::SENT_PREEMPT:
      switch (msg->type) {
        case easynav_interfaces::msg::NavigationControl::FEEDBACK:
          RCLCPP_DEBUG(node_->get_logger(), "Getting navigation feedback");
          last_feedback_ = *msg;
          break;
        case easynav_interfaces::msg::NavigationControl::ACCEPT:
          RCLCPP_DEBUG(node_->get_logger(), "Goal preemption accepted. Navigating");
          state_ = State::ACCEPTED_AND_NAVIGATING;
          break;
        case easynav_interfaces::msg::NavigationControl::REJECT:
          RCLCPP_ERROR(node_->get_logger(), "Rejected navigation preempt goal");
          state_ = State::ACCEPTED_AND_NAVIGATING;
          break;
        case easynav_interfaces::msg::NavigationControl::ERROR:
          RCLCPP_ERROR(node_->get_logger(), "Error in preempting navigation");
          state_ = State::ERROR;
          break;
        default:
          RCLCPP_ERROR(
            node_->get_logger(), "State SENT_PREEMPT; Unexpected message: %d: %s",
            msg->type, msg->status_message.c_str());
          state_ = State::ERROR;
      }
      break;
    case State::ACCEPTED_AND_NAVIGATING:
      switch (msg->type) {
        case easynav_interfaces::msg::NavigationControl::FEEDBACK:
          RCLCPP_DEBUG(node_->get_logger(), "Getting navigation feedback");
          last_feedback_ = *msg;
          break;
        case easynav_interfaces::msg::NavigationControl::FINISHED:
          RCLCPP_INFO(node_->get_logger(), "Navigation succesfully finished");
          last_result_ = *msg;
          state_ = State::NAVIGATION_FINISHED;
          break;
        case easynav_interfaces::msg::NavigationControl::FAILED:
          RCLCPP_ERROR(
            node_->get_logger(), "Navigation with error finished: %s",
            msg->status_message.c_str());
          last_result_ = *msg;
          state_ = State::NAVIGATION_FAILED;
          break;
        case easynav_interfaces::msg::NavigationControl::CANCELLED:
          RCLCPP_INFO(node_->get_logger(), "Navigation cancelled");
          last_result_ = *msg;
          state_ = State::NAVIGATION_CANCELLED;
          break;
        default:
          RCLCPP_ERROR(
            node_->get_logger(), "State ACCEPTED_AND_NAVIGATING; Unexpected message: %d: %s",
            msg->type, msg->status_message.c_str());
          last_result_ = *msg;
          state_ = State::ERROR;
      }
      break;
    default:
      RCLCPP_ERROR(
        node_->get_logger(), "State not managed: %d:",
        static_cast<int>(state_));
      state_ = State::ERROR;
  }

  last_control_ = std::move(msg);
}

const easynav_interfaces::msg::NavigationControl &
GoalManagerClient::get_last_control() const
{
  return *last_control_;
}

const easynav_interfaces::msg::NavigationControl &
GoalManagerClient::get_feedback() const
{
  return last_feedback_;
}

const easynav_interfaces::msg::NavigationControl &
GoalManagerClient::get_result() const
{
  return last_result_;
}

void
GoalManagerClient::send_goal(const geometry_msgs::msg::PoseStamped & goal)
{
  RCLCPP_DEBUG(node_->get_logger(), "Sending navigation goal");

  nav_msgs::msg::Goals msg;
  msg.header = goal.header;
  msg.goals.push_back(goal);

  send_goals(msg);
}

void
GoalManagerClient::send_goals(const nav_msgs::msg::Goals & goals)
{
  if (last_control_ == nullptr) {
    last_control_ = std::make_unique<easynav_interfaces::msg::NavigationControl>();
  }

  if (goals.goals.empty()) {
    RCLCPP_ERROR(node_->get_logger(), "Trying to command empty goals");
    return;
  }

  easynav_interfaces::msg::NavigationControl msg;

  switch(state_) {
    case State::IDLE:
      state_ = State::SENT_GOAL;
      msg.type = easynav_interfaces::msg::NavigationControl::REQUEST;
      break;
    case State::ACCEPTED_AND_NAVIGATING:
      state_ = State::SENT_PREEMPT;
      msg.type = easynav_interfaces::msg::NavigationControl::REQUEST;
      break;
    default:
      RCLCPP_ERROR(node_->get_logger(), "Trying to send new goals in state %d. Ignoring",
      static_cast<int>(state_));
      return;
  }

  msg.header = goals.header;
  msg.user_id = id_;
  msg.seq = last_control_->seq + 1;
  msg.goals = goals;

  control_pub_->publish(msg);
}

void
GoalManagerClient::cancel()
{
  RCLCPP_DEBUG(node_->get_logger(), "Sending nevigation cancelation");

  if (state_ != State::ACCEPTED_AND_NAVIGATING) {
    RCLCPP_ERROR(
      node_->get_logger(),
      "Triying to cancel a non-active navigation (state %d)",
       static_cast<int>(state_));
    return;
  }

  easynav_interfaces::msg::NavigationControl msg;
  msg.type = easynav_interfaces::msg::NavigationControl::CANCEL;
  msg.header = last_control_->header;
  msg.user_id = id_;
  msg.seq = last_control_->seq + 1;


  RCLCPP_DEBUG(node_->get_logger(), "Navigation cancelation sent");
  control_pub_->publish(msg);
}

void
GoalManagerClient::reset()
{
  if (state_ == State::NAVIGATION_FINISHED || state_ == State::NAVIGATION_REJECTED ||
    state_ == State::NAVIGATION_FAILED || state_ == State::NAVIGATION_CANCELLED ||
    state_ == State::ERROR)
  {
    state_ = State::IDLE;
  } else {
    RCLCPP_ERROR(
      node_->get_logger(),
      "Triying to reset navigation in a a non-finished navigation state %d",
       static_cast<int>(state_));
  }
}

}  // namespace easynav
